packages/search/src/search.tsx

```diff
@@ -0,0 +1,417 @@
+import { close, search, Icon } from '@wordpress/icons';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Style dependencies
+ */
+import './style.scss';
+
+/**
+ * Internal variables
+ */
+const SEARCH_DEBOUNCE_MS = 300;
+
+const keyListener = (
+	methodToCall: ( e: MouseEvent< HTMLDivElement > | KeyboardEvent< HTMLDivElement > ) => void
+) => ( event: KeyboardEvent< HTMLDivElement > ) => {
+	switch ( event.key ) {
+		case ' ':
+		case 'Enter':
+			methodToCall( event );
+			break;
+	}
+};
+
+type Props = {
+	autoFocus: boolean;
+	className?: string;
+	compact: boolean;
+	delaySearch: boolean;
+	delayTimeout: number;
+	describedBy?: string;
+	dir?: 'ltr' | 'rlt';
+	disableAutocorrect: boolean;
+	disabled: boolean;
+	fitsContainer: boolean;
+	hideClose: boolean;
+	hideOpenIcon: boolean;
+	initialValue?: string;
+	inputLabel?: string;
+	isOpen: boolean;
+	maxLength?: number;
+	minLength?: number;
+	onBlur?: ( event: MouseEvent< HTMLInputElement > ) => void;
+	onClick?: () => void;
+	onKeyDown?: ( event: KeyboardEvent< HTMLInputElement > ) => void;
+	onSearch: ( search: string ) => void;
+	onSearchChange: ( search: string ) => void;
+	onSearchOpen: ( event?: MouseEvent< HTMLDivElement > | KeyboardEvent< HTMLDivElement > ) => void;
+	onSearchClose: ( event: MouseEvent< HTMLDivElement > | KeyboardEvent< HTMLDivElement > ) => void;
+	overlayStyling?: ( search: string ) => React.ReactNode;
+	placeholder?: string;
+	pinned: boolean;
+	searching: boolean;
+	value?: string;
+};
+
+type State = {
+	keyword: string;
+	isOpen: boolean;
+	hasFocus: boolean;
+};
+
+class Search extends React.Component< Props, State > {
+	static defaultProps = {
+		autoFocus: false,
+		compact: false,
+		delaySearch: false,
+		delayTimeout: SEARCH_DEBOUNCE_MS,
+		describedBy: null,
+		dir: undefined,
+		disableAutocorrect: false,
+		disabled: false,
+		fitsContainer: false,
+		hideClose: false,
+		hideOpenIcon: false,
+		isOpen: false,
+		onClick: noop,
+		onKeyDown: noop,
+		onSearchChange: noop,
+		onSearchOpen: noop,
+		onSearchClose: noop,
+		//undefined value for overlayStyling is an optimization that will
+		//disable overlay scrolling calculation when no overlay is provided.
+		overlayStyling: undefined,
+		pinned: false,
+		searching: false,
+	};
+
+	instanceId = uniqueId();
+	searchInput = React.createRef< HTMLInputElement >();
+	openIcon = React.createRef< HTMLDivElement >();
+	overlay = React.createRef< HTMLDivElement >();
+
+	// debounced `onSearch` will have a `cancel` function
+	onSearch: ( ( search: string ) => void ) & { cancel?: () => void } = this.props.delaySearch
+		? debounce( this.props.onSearch, this.props.delayTimeout )
+		: this.props.onSearch;
+
+	state = {
+		keyword: this.props.initialValue ?? '',
+		isOpen: this.props.isOpen,
+		hasFocus: this.props.autoFocus,
+	};
+
+	UNSAFE_componentWillReceiveProps( nextProps: Props ) {
+		if ( this.props.isOpen !== nextProps.isOpen ) {
```

Comment:
This feels a bit flawed. The component state can be changed by user intent but could be overridden by another prop change, feels a bit off. I'd expect `isOpen` to be used only for the initial state. Is there a real use case where this behavior is necessary?

Reply:
> These cases would maybe be covered if the search component auto-closed when it loses focus -- means that another UI element is now active.

For this to work we'd need to always auto-focus. I'm not opposed to that given that a similar component in the new P2 follows this behavior but my understanding is that auto-focusing is undesirable for accessibility reasons. Maybe that doesn't apply for when you're "auto-focusing" on a new element after activating a button that makes the new element appear?

Reply:
As far as I understand, auto-focusing is always intrusive from an accessibility point of view. I might be wrong, of course - @diegohaz would know best.

Reply:
Auto-focus is okay if followed by a user action that makes sense. For example, clicking on a `<label>` should focus on the input; opening a dialog should move focus into it; deleting the currently focused row in a table could move focus somewhere else (like the next or previous rows).

In this component, I would expect that the input would be focused after clicking on the search icon.